code/game/machinery/spaceheater.dm

```diff
@@ -82,85 +82,76 @@
 	else
-		on = !on
-		user.visible_message("<span class='notice'>[user] switches [on ? "on" : "off"] [src].</span>","<span class='notice'>You switch [on ? "on" : "off"] [src].</span>")
-		update_icon()
-	return
-
-
-/obj/machinery/space_heater/Topic(href, href_list)
-	if(..())
-		return 1
-	if((in_range(src, usr) && isturf(src.loc)) || (issilicon(usr)))
-		usr.set_machine(src)
-
-		switch(href_list["op"])
-
-			if("temp")
-				var/value = text2num(href_list["val"])
-
-				// limit to 20-90 degC
-				set_temperature = dd_range(0, 90, set_temperature + value)
-
-			if("cellremove")
-				if(open && cell && !usr.get_active_hand())
-					cell.update_icon()
-					usr.put_in_hands(cell)
-					cell.add_fingerprint(usr)
-					cell = null
-					usr.visible_message("<span class='notice'>[usr] removes the power cell from [src].</span>", "<span class='notice'>You remove the power cell from [src].</span>")
-
-
-			if("cellinstall")
-				if(open && !cell)
-					var/obj/item/stock_parts/cell/C = usr.get_active_hand()
-					if(istype(C))
-						usr.drop_item()
-						cell = C
-						C.loc = src
-						C.add_fingerprint(usr)
+		toggle_power(user)
+
+/obj/machinery/space_heater/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "SpaceHeater", "Space Heater Internals", 360, 250)
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+/obj/machinery/space_heater/ui_data(mob/user)
+	var/list/data = list()
+	data["Powercell"] = cell ? cell : FALSE
+	data["CellPercent"] = cell ? round(cell.percent(),1) : 0
+	data["Temp"] = set_temperature
+	data["on"] = on
+	data["open"] = open
+	return data
+
+/obj/machinery/space_heater/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	if(..() || !open)
+		return
 
-						usr.visible_message("<span class='notice'>[usr] inserts a power cell into [src].</span>", "<span class='notice'>You insert the power cell into [src].</span>")
+	. = TRUE
 
-		updateDialog()
-	else
-		usr << browse(null, "window=spaceheater")
-		usr.unset_machine()
-	return
+	switch(action)
+		if("change_temp")
+			set_temperature = dd_range(0, 90, text2num(params["change_temp"]))
+		if("add_cell")
+			if(!open || cell)
+				return
+			var/obj/item/stock_parts/cell/C = usr.get_active_hand()
+			if(!istype(C))
+				return
+			usr.drop_item()
+			cell = C
+			C.forceMove(src)
+			C.add_fingerprint(usr)
+
+			usr.visible_message("<span class='notice'>[usr] inserts a power cell into [src].</span>", "<span class='notice'>You insert the power cell into [src].</span>")
+		if("remove_cell")
+			if(!open || !cell || usr.get_active_hand())
+				return
+			on = FALSE
+			cell.update_icon()
+			usr.put_in_hands(cell)
+			cell.add_fingerprint(usr)
+			cell = null
+			usr.visible_message("<span class='notice'>[usr] removes the power cell from [src].</span>", "<span class='notice'>You remove the power cell from [src].</span>")
```

Comment:
Will be good to move `remove_cell` and `add_cell` to separate procs, for readability